src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt

```diff
@@ -226,101 +202,220 @@ class IncomingPaymentHandler(val nodeParams: NodeParams, val db: IncomingPayment
+                                        }
+                                        acceptPayment(incomingPayment, receivedWith, actions)
+                                    }
                                 }
                             }
+                        }
+                    }
+                }
+            }
+        }
+    }
 
-                            pending.remove(paymentPart.paymentHash)
-                            val received = IncomingPayment.Received(receivedWith = receivedWith)
-                            if (incomingPayment.origin is IncomingPayment.Origin.Offer) {
-                                // We didn't store the Bolt 12 invoice in our DB when receiving the invoice_request (to protect against DoS).
-                                // We need to create the DB entry now otherwise the payment won't be recorded.
-                                db.addIncomingPayment(incomingPayment.preimage, incomingPayment.origin)
-                            }
-                            db.receivePayment(paymentPart.paymentHash, received.receivedWith)
-                            nodeParams._nodeEvents.emit(PaymentEvents.PaymentReceived(paymentPart.paymentHash, received.receivedWith))
-                            return ProcessAddResult.Accepted(actions, incomingPayment.copy(received = received), received)
+    private suspend fun acceptPayment(incomingPayment: IncomingPayment, receivedWith: List<IncomingPayment.ReceivedWith>, actions: List<PeerCommand>): ProcessAddResult.Accepted {
+        pending.remove(incomingPayment.paymentHash)
+        if (incomingPayment.origin is IncomingPayment.Origin.Offer) {
+            // We didn't store the Bolt 12 invoice in our DB when receiving the invoice_request (to protect against DoS).
+            // We need to create the DB entry now otherwise the payment won't be recorded.
+            db.addIncomingPayment(incomingPayment.preimage, incomingPayment.origin)
+        }
+        db.receivePayment(incomingPayment.paymentHash, receivedWith)
+        nodeParams._nodeEvents.emit(PaymentEvents.PaymentReceived(incomingPayment.paymentHash, receivedWith))
+        val received = IncomingPayment.Received(receivedWith)
+        return ProcessAddResult.Accepted(actions, incomingPayment.copy(received = received), received)
+    }
+
+    private fun rejectPayment(payment: PendingPayment, incomingPayment: IncomingPayment, failure: FailureMessage): ProcessAddResult.Rejected {
+        pending.remove(incomingPayment.paymentHash)
+        val actions = payment.parts.map { part ->
+            when (part) {
+                is HtlcPart -> actionForFailureMessage(failure, part.htlc)
+                is WillAddHtlcPart -> actionForWillAddHtlcFailure(nodeParams.nodePrivateKey, failure, part.htlc)
+            }
+        }
+        return ProcessAddResult.Rejected(actions, incomingPayment)
+    }
+
+    private fun validateOnTheFlyFundingRate(
```

Comment:
Can you explain again why we need to call `maybeReject` two times? 🙏

In the legacy fee-credit branch, [the semantics of `maybeReject` has been changed](https://github.com/ACINQ/lightning-kmp/blob/fee-credit-11/src/commonMain/kotlin/fr/acinq/lightning/payment/LiquidityPolicy.kt), making it return a `Decision` instead of a `Rejection`. Would it help the factorization?

Reply:
Fair enough, we can always review later.